### PR TITLE
fix(vuln): GitLab report template

### LINKED ISSUE #1598: Integration with GitLab will stop working in GitLab 15.0
[Container scanning schemas below 14.0.0](https://docs.gitlab.com/ee/update/deprecations#container-scanning-schemas-below-1400) have been deprecated.

[`blob/main/contrib/gitlab.tpl:3`](https://github.com/aquasecurity/trivy/blob/main/contrib/gitlab.tpl#L3) is using a deprecated version:

>   "version": "2.3",

The latest version of the schema is [14.1.0](https://gitlab.com/gitlab-org/gitlab/-/tree/master/ee/lib/ee/gitlab/ci/parsers/security/validators/schemas/14.1.0). There are other [schemas available](https://gitlab.com/gitlab-org/gitlab/-/tree/master/ee/lib/ee/gitlab/ci/parsers/security/validators/schemas); any version from 14.0.0 is supported.

From GitLab 15.0, any reports submitted using an unsupported version will be rejected.

### PR DESCRIPTION
**Pull request by thiago-gitlab**
## Description

- Upgrade to schema 14.0.6 (https://gitlab.com/gitlab-org/security-products/security-report-schemas/-/blob/v14.0.6/dist/container-scanning-report-format.json).
- Drop unsupported `confidence` property. Currently optional and will be removed by GitLab in schema 15-0-0.

I've generated a test report using:

```
trivy image --exit-code 0 --security-checks vuln --format template --template "@contrib/gitlab.tpl" -o gl-container-scanning-report.json webgoat/webgoat-8.0:latest
```

And I've [validated locally](https://docs.gitlab.com/ee/development/integrations/secure.html#validate-locally):
```
Thiagos-MacBook-Pro:tmp thiago$ ruby validate.rb container-scanning-report-format_14-0-6.json ~/repos/aquasecurity/trivy/gl-container-scanning-report.json
property '/vulnerabilities/8/identifiers/0/url' does not match format: uri
property '/vulnerabilities/11/identifiers/0/url' does not match format: uri
property '/vulnerabilities/140/links/4/url' does not match format: uri
property '/vulnerabilities/140/links/5/url' does not match format: uri
property '/vulnerabilities/140/links/6/url' does not match format: uri
property '/vulnerabilities/147/links/6/url' does not match format: uri
property '/vulnerabilities/147/links/7/url' does not match format: uri
property '/vulnerabilities/148/links/21/url' does not match format: uri
property '/vulnerabilities/148/links/22/url' does not match format: uri
property '/vulnerabilities/177/identifiers/0/url' does not match format: uri
property '/vulnerabilities/267/identifiers/0/url' does not match format: uri
property '/vulnerabilities/317/links/22/url' does not match format: uri
property '/vulnerabilities/331/links/4/url' does not match format: uri
property '/vulnerabilities/354/links/4/url' does not match format: uri
property '/vulnerabilities/443/links/22/url' does not match format: uri
property '/vulnerabilities/469/identifiers/0/url' does not match format: uri
property '/vulnerabilities/470/identifiers/0/url' does not match format: uri
property '/vulnerabilities/471/identifiers/0/url' does not match format: uri
property '/vulnerabilities/472/identifiers/0/url' does not match format: uri
property '/vulnerabilities/473/identifiers/0/url' does not match format: uri
property '/vulnerabilities/474/identifiers/0/url' does not match format: uri
property '/vulnerabilities/486/links/3/url' does not match format: uri
```

The errors above are because:

1. Some Debian advisories are missing the `url`
   ```json
         "identifiers": [
           {
             "type": "cve",
             "name": "DLA-2424-1",
             "value": "DLA-2424-1",
             "url": ""
           }
         ],
   ```
2. Malformed URLs. E.g.: `"https://gitlab.freedesktop.org/dbus/dbus/-/commit/2b7948ef907669e844b52c4fa2268d6e3162a70c (dbus-1.13.18)"`. 

These invalid entries display as errors in the GitLab pipeline security tab, and are ignored when ingesting the data for vulnerability reports.

Finally, I was going to add a `scan` section, since this will be required soon. Unfortunately my Golang is lacking, and I can't seem to find an existing variable that I can use in the template to output the Trivy version. E.g.:

```json
  "scan": {
    "scanner": {
      "id": "trivy",
      "name": "Trivy",
      "url": "https://github.com/aquasecurity/trivy/",
      "vendor": {
        "name": "Aqua Security"
      },
      "version": "0.32.2"
    }
  },
```

## Related issues
- Close #1598

## Related PRs
- [ ] https://github.com/aquasecurity/trivy/pull/2153

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [ ] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: contrib/gitlab.tpl [hunk 1]
```diff
@@ -1,10 +1,11 @@
 {{- /* Template based on https://docs.gitlab.com/ee/user/application_security/container_scanning/#reports-json-format */ -}}
 {
-  "version": "2.3",
+  "version": "14.0.6",
   "vulnerabilities": [
   {{- $t_first := true }}
   {{- range . }}
   {{- $target := .Target }}
+    {{ $image := $target | regexFind "^[^:\\s]+" }}
```
Comment (thiago-gitlab) [inline contrib/gitlab.tpl:8]: Remove the OS from the image.

While version `14.0.6` of the schema doesn't enforce the format, version `14.1.x` and later versions do. So I opted for using the string up to a space; e.g. `webgoat/webgoat-8.0:latest (debian 9.6)` becomes `webgoat/webgoat-8.0:latest`.

See also https://github.com/aquasecurity/trivy/issues/332.

File: contrib/gitlab.tpl [hunk 2]
```diff
@@ -1,10 +1,11 @@
 {{- /* Template based on https://docs.gitlab.com/ee/user/application_security/container_scanning/#reports-json-format */ -}}
 {
-  "version": "2.3",
+  "version": "14.0.6",
   "vulnerabilities": [
   {{- $t_first := true }}
   {{- range . }}
   {{- $target := .Target }}
+    {{ $image := $target | regexFind "[^\\s]+" }}
```
Comment (thiago-gitlab) [inline contrib/gitlab.tpl:8]: Not exactly a valid docker image string either, but `14.0.6` doesn't validate the pattern, so we can be less strict here.

**Comment by thiago-gitlab** [comment]
@knqyf263, there's an integration test failing. Unfortunately it doesn't show me a diff of what it's expecting, and I'm not sure this can be run locally.

Do you have a suggestion?

```
    --- FAIL: TestClientServerWithFormat/alpine_3.10_with_gitlab_template (0.13s)
```

**Comment by knqyf263** [comment]
The failing test shows the following diff.

```
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -3,2 +3,3 @@
        	            	   "vulnerabilities": [
        	            	+    
        	            	     {
```

Looks like there is a new line, but I didn't see such a change. @DmitriyLewen Could you help us?

File: contrib/gitlab.tpl (hunk 2 above)
Comment (DmitriyLewen) [inline contrib/gitlab.tpl:8]:
```suggestion
    {{- $image := $target | regexFind "[^\\s]+" -}}
```

@knqyf263 , @thiago-gitlab this should help with the error in `TestClientServerWithFormat`

Comment (thiago-gitlab) [reply contrib/gitlab.tpl:8]: Many thanks, @DmitriyLewen. I'll fix that today 🙇🏽

**Comment by thiago-gitlab** [comment]
Thanks for the help, @knqyf263 and @DmitriyLewen. Checks are passing. Ready for another round of review 🏓

**Review by knqyf263 — APPROVED** [review]

**Comment by knqyf263** [comment]
Thanks a lot!

**Pull request merged**